wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,386 @@
+	but they are disabled to modify. */
+public interface DamageListener
+	function listen(Damage d)
+
+let damageModifiers = new LinkedList<DamageModifier>
+
+let damageListeners = new LinkedList<DamageListener>
+
+public function addDamageModifier(DamageModifier cb) returns DamageModifier
+	damageModifiers.add(cb)
+	return cb
+
+public function removeDamageModifier(DamageModifier cb)
+	damageModifiers.remove(cb)
+
+public function addDamageListener(DamageListener cb) returns DamageListener
+	damageListeners.add(cb)
+	return cb
+
+public function removeDamageListener(DamageListener cb)
+	damageListeners.remove(cb)
+
+// The wrapper class
+
+enum DamageState
+	CREATED // Initial state after creation
+	FIRING // Damage.deal() is called
+	ACQUIRED // Damage is acquired by damage event handler
+	MODIFYING // Damage.modify() is called
+	LISTENING // Last processing stage; modifying is forbidden
+	FIRED // Damage instance is completely processed by system
+
+public function DamageState.toString() returns string
+	switch this
+		case CREATED
+			return "created"
+		case FIRING
+			return "firing"
+		case ACQUIRED
+			return "acquired"
+		case MODIFYING
+			return "modifying"
+		case LISTENING
+			return "listening"
+		case FIRED
+			return "fired"
+
+	return "error"
+	
+public class Damage
+
+	// Event info
+
+	/** Custom field which can be pinned damage event */
+	integer data
+
+	private unit source
+	private unit target
+	private bool ranged
+	private bool attack
+	private attacktype attackType
+	private damagetype damageType
+	private weapontype weaponType
+
+	/** What amount of damage was originally convieved */
+	private real initialAmount
+	/** What amount of damage was convieved after modification*/
+	private real amount
+	/** Final amount of damage event */
+	protected real actualAmount = 0
+	
+	/** Holds info about damage event state */
+	protected DamageState state = DamageState.CREATED
+	/** It is set to true if the damage, when it was dealt, killed the target */
+	protected bool killingBlow = false
+
+	function getSource() returns unit
+		return source
+	function getTarget() returns unit
+		return target
+	function getRanged() returns bool
+		return ranged
+	function getAttack() returns bool
+		return attack
+	function getAttackType() returns attacktype
+		return attackType
+	function getDamageType() returns damagetype
+		return damageType
+	function getWeaponType() returns weapontype
+		return weaponType
+	function getState() returns DamageState
+		return state
+
+	/** If the damage is modified by spell bracers ability. */
+	function isSpell() returns bool
+		return attackType == ATTACK_TYPE_NORMAL
+
+	/** Returns true if the damage is going to kill or already have killed target */
+	function isKillingBlow() returns bool
+		return (state == DamageState.FIRED and killingBlow) or (state != DamageState.FIRED and target.getHP() - .406 < amount)
+
+	function getInitial() returns real
```

Comment:
Those three have different behavior, depending  on your needs you would pick a different one.

Reply:
Three getters are fine, but being able to use them incorrectly is not.
getActual only working in listeners, and being 0 without and user feedback or errors outside of listeners, is not acceptable.

Reply:
I agree with Frotty. I'm still for simplification of this interface. I think it would make more sense to have `getAmount` return the amount that makes the most sense, given the state of the Damage object.

Reply:
But there are cases where you could want to know all three amounts at once. Or at least it would be bad to deny accessing information. Maybe `getActual()` should return `getAmount()` before listenening stage... I think having a warning in debug stream and amount returned would be a fitting solution.